### ISSUE TITLE
Geo-blocking fails on multiple users

### ISSUE THREAD

**Issue by Thorium** [edgarfgp/GeoblockingMiddleware#1]
Current version of `updateMemoryCache` crash the process if used more than 1 user at the same time:
```
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at updateMemoryCache(String ipAddress, GeoblockStatus status) 
```

Here is a small sample code to test, that could be added to a unit-test:
```fsharp
// <code taken from geoblocking>
open System
open System.Collections.Generic

type IpApiModel =
    { status: string
      countryCode: string }

type GeoblockStatus =
    { IsBlocked: bool
      LastAccessTime: DateTime }

let [<Literal>] MAX_IP_IN_CACHE = 100
let lastIpAddresses = Dictionary<string, GeoblockStatus>()

let updateMemoryCache (ipAddress: string) (status: GeoblockStatus) =
    if lastIpAddresses.ContainsKey(ipAddress) then
        lastIpAddresses[ipAddress] <- status

    else
        if lastIpAddresses.Count > MAX_IP_IN_CACHE then
            let oldestIp = lastIpAddresses |> Seq.sortBy(fun kvp -> kvp.Value.LastAccessTime) |> Seq.head
            lastIpAddresses.Remove(oldestIp.Key) |> ignore

        lastIpAddresses.Add(ipAddress, status)
// </code taken from geoblocking>

// <unit-test>
open System
open System.Threading
let rnd = new Random()

let taskarray = 
    [1u..100u] |> Seq.map(fun itm ->
        let rndIp() =
            ("." , Array.unfold (fun x -> 
                if x < 4 then Some (rnd.Next(0, 255).ToString(), x+1)
                else None) 0) |> String.Join

        let t1 = Tasks.Task.Run(fun () ->
            Console.WriteLine Thread.CurrentThread.ManagedThreadId
            (rndIp(),{IsBlocked=false; LastAccessTime=DateTime.Now}) 
                ||> updateMemoryCache
        )
        let t2 = Tasks.Task.Run(fun () ->
            let ip = rndIp()
            let getip() =
                (ip,{IsBlocked=false; LastAccessTime=DateTime.Now}) 
                    ||> updateMemoryCache
            Console.WriteLine Thread.CurrentThread.ManagedThreadId
            getip(); getip() // add and get
        )
        Tasks.Task.WhenAll [|t1; t2|]
        //ctx.ClearUpdates() |> ignore
    ) |> Seq.toArray |> Tasks.Task.WaitAll

printfn "we never reach this far although we should"
//let ack = Console.ReadLine()
```

**Comment by Thorium** [comment]
`lastIpAddresses` type should probably be `ConcurrentDictionary<string, Task<Lazy<GeoblockStatus>>>` or `ConcurrentQueue<Task<Lazy<Tuple<string*GeoblockStatus>>>>` or similar...

**Comment by Thorium** [comment]
This was fixed by commit 939f77a2fe5ca0e588a20b76e74847b36960ad30
